083-stretch-cluster.md

```diff
@@ -0,0 +1,199 @@
+This necessitates careful consideration of network architecture and associated costs.
+
+- **Latency Requirements**: The stretch Kafka cluster is best suited for environments with low-latency network connections between the Kubernetes clusters.High latency can adversely affect the performance and synchronization of Kafka nodes, potentially leading to delays or errors in replication and client communication.
+Defining the minimal acceptable latency between clusters is crucial to ensure optimal performance.
+
+## Proposal
+
+This proposal seeks to enhance the Strimzi Kafka operator to support stretch Kafka clusters, distributing brokers and controllers across multiple Kubernetes clusters.
+The intent is to focus on high-availability of the data plane.The proposal outlines high-level topology and design concepts for such deployments, with a plan to incrementally include finer design and implementation details for various aspects.
+
+### Prerequisites
+
+- **Multiple Kubernetes Clusters**: Stretch Kafka clusters will require multiple Kubernetes clusters.
+Ideally, an odd number of clusters (at least three) is needed to maintain quorum in the event of a cluster outage.
+
+- **Low Latency**: Kafka clusters should be deployed in environments that allow low-latency communication between Kafka brokers and controllers.
+Stretch Kafka clusters should be deployed in environments such as data centers or availability zones within a single region, and not across distant regions where high latency could impair performance.
+
+- **KRaft**: As Kafka and Strimzi transition towards KRaft-based clusters, this proposal focuses exclusively on enabling stretch deployments for KRaft-based Kafka clusters.
+While Zookeeper-based deployments are still supported, they are outside the scope of this proposal.
+
+### Design
+
+The cluster operator will be deployed in all Kubernetes clusters and will manage Kafka brokers/controllers running on that cluster.
+One Kubernetes cluster will act as the control point for defining custom resources (Kafka, KafkaNodePool) required for stretch Kafka cluster.
+The KafkaNodePool custom resource will be extended to include information about a Kubernetes cluster where the pool should be deployed.
+The cluster operator will create necessary resources (StrimziPodSets, services etc.) on the target clusters specified within the KafkaNodePool resource.
+
+This approach will allow users to specify/manage the definition of stretch Kafka cluster in a single location.
+The operators will then create necessary resources in target Kubernetes clusters, which can then be reconciled/managed by operators on those clusters.
+
+### Reconciling Kafka and KafkaNodePool resources
+![Reconciling Kafka and KafkaNodePool resources](./images/083-reconciling-kafka-knp.png)
```

Comment:
From this picture, my understanding is that there is just one cluster operator running in one Kube cluster where you deploy the custom resources ... while the other one has more operators (one for each cluster). I think just one picture would be enough ... AFAIU you envisage the other operators just to handle SPS but not other custom resources? or can they be used to handle local clusters as well?
I was wondering if just one operator is enough and it can reconcile SPS on other Kube clusters just talking to the right remote Kube API.

Reply:
The idea is that the operators running in the other Kubernetes clusters (not the central cluster) will only reconcile StrimziPodSets. This design ensures resilience in the event of a failure in the central Kubernetes cluster. If the central cluster crashes unexpectedly, the clients can still be served because both the Cluster Operator and Kafka pods will continue to run in the other clusters.

The Cluster Operator in each respective cluster will reconcile StrimziPodSets, ensuring that the broker and controller pods in that specific cluster remain active and continue serving clients. Once the central cluster is restored and becomes available, the other clusters can rejoin it seamlessly. This approach ensures zero downtime, even in scenarios where the central Kubernetes cluster experiences an outage.